src/ServiceManager.php

```diff
@@ -968,4 +958,33 @@ private function abstractFactoryCanCreate(string $name): bool
 
         return false;
     }
+
+    /**
+     * @psalm-param mixed $delegatorFactory
+     * @psalm-assert callable(ContainerInterface,string,callable,array<mixed>):object $delegatorFactory
+     */
+    private function assertCallableDelegatorFactory($delegatorFactory): void
+    {
+        if ($delegatorFactory instanceof Factory\DelegatorFactoryInterface) {
+            return;
+        }
+
+        if (! is_callable($delegatorFactory)) {
+            if (is_string($delegatorFactory)) {
+                throw new ServiceNotCreatedException(sprintf(
+                    'An invalid delegator factory was registered; resolved to class or function "%s"'
+                    . ' which does not exist; please provide a valid function name or class name resolving'
+                    . ' to an implementation of %s',
+                    $delegatorFactory,
+                    DelegatorFactoryInterface::class
+                ));
+            }
+
+            throw new ServiceNotCreatedException(sprintf(
+                'A non-callable delegator, "%s", was provided; expected a callable or instance of "%s"',
+                is_object($delegatorFactory) ? get_class($delegatorFactory) : gettype($delegatorFactory),
+                DelegatorFactoryInterface::class
+            ));
+        }
+    }
```

Comment:
```suggestion
        if ($delegatorFactory instanceof Factory\DelegatorFactoryInterface
            || is_callable($delegatorFactory)
        ) {
            return;
        }

        if (is_string($delegatorFactory)) {
            throw new ServiceNotCreatedException(sprintf(
                'An invalid delegator factory was registered; resolved to class or function "%s"'
                . ' which does not exist; please provide a valid function name or class name resolving'
                . ' to an implementation of %s',
                $delegatorFactory,
                DelegatorFactoryInterface::class
            ));
        }

      throw new ServiceNotCreatedException(sprintf(
            'A non-callable delegator, "%s", was provided; expected a callable or instance of "%s"',
            is_object($delegatorFactory) ? get_class($delegatorFactory) : gettype($delegatorFactory),
            DelegatorFactoryInterface::class
        ));
    }
```

Basically, return early as soon as we know it's valid, and otherwise just vary the exception based on whether or not it was passed as a string.